### DISCUSSION TITLE
Is there a way to disable trailing whitespace character (&#x20;)?

### DISCUSSION THREAD

**Discussion by dannyadkins** [syntax-tree/unist#122] [Q&A] [answered]
When you type and leave a space at the end of e.g. a paragraph block, it is replaced with "&#x20;" placeholder.

I can't find exactly where in the code this is getting injected.

Is there a way to disable this? We would like to get visually "clean" text outputted from our editor without this special character.

Thanks for any help!

**Comment by ChristianMurphy** [comment]
Welcome @dannyadkins! 👋 

Could you frame your question in a but more detail? See https://github.com/syntax-tree/.github/blob/main/support.md#support for recommendations.

It's unclear:
1. what language you are referring to (markdown, html, xml, etc)?
2. Which library you are using?
3. What the version of that library are you using?
4. What input document you are you trying?
5. What is the output you are expecting?
6. What is unexpected about a space character that was added to the document by the author, remaining in the document?

**Reply by dannyadkins** [reply]
my bad!! @ChristianMurphy 

I'm using markdown with this library: https://github.com/syntax-tree/mdast-util-to-markdown/tree/main/lib most recent version. 

For example, if I input "some text " with a trailing space, it gets encoded to markdown: "some text&#x20;". If I input "some text   " (with many trailing spaces), only the last one gets encoded as that character (output would be "some text  &#x20;"). 

You can see this behavior in https://milkdown.dev/playground which implements this library in the background.

It's not *necessarily* unexpected, but it is inconvenient for our use case and we are having to manually parse it out.

**Reply by ChristianMurphy** [reply, marked as answer]
> You can see this behavior in https://milkdown.dev/playground which implements this library in the background.

Milkdown adds its own transforms.
If you use `mdast-util-to-markdown` directly through `remark` no character is added.
https://codesandbox.io/p/devbox/crazy-fermi-lzjy7p?file=%2Fsrc%2Fmain.ts%3A24%2C1&workspaceId=d845f959-45b3-427c-9d19-50c1c90a9035

This behavior is being set by Milkdown through a custom plugin https://github.com/Milkdown/milkdown/blob/063238906971aceed1bdafd6b5a634433dd4da3b/packages/preset-commonmark/src/plugin/remark-line-break.ts#L7C1-L9

**Reply by ChristianMurphy** [reply]
Beyond the non-standard behavior added by Milkdown, the way `mdast-util-to-markdown` is handling their custom input is necessary.

this represents a line break
```
some text  
```

this represents two spaces
```
some text &#x20;
```

that is how markdown works. https://spec.commonmark.org/0.30/#hard-line-breaks

**Reply by dannyadkins** [reply]
wow thank you so much, completely missed that -- really appreciate it!
